From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Praschl <praschlpaul@g-p.at>
Date: Tue, 4 Aug 2026 23:31:17 +0200
Subject: [PATCH 112/117] Input: apple_z2: Release missing D11 contacts

The D11 controller omits released contacts from subsequent reports.
Without INPUT_MT_DROP_UNUSED, slots that no longer appear remain active
and produce stuck points after multi-finger releases.

Enable automatic release only for D11 and leave other variants
unchanged.

Signed-off-by: Paul Praschl <praschlpaul@g-p.at>

Orchard-replay-of: 5a4b15dfdf3a9526e97d07ea4af5756c47c15ba4
---
 drivers/input/touchscreen/apple_z2.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/input/touchscreen/apple_z2.c b/drivers/input/touchscreen/apple_z2.c
index ab5b02a50..ec84de96f 100644
--- a/drivers/input/touchscreen/apple_z2.c
+++ b/drivers/input/touchscreen/apple_z2.c
@@ -2226,6 +2226,7 @@ static int apple_z2_probe(struct spi_device *spi)
 	struct device *dev = &spi->dev;
 	const struct apple_z2_chip_info *info;
 	struct apple_z2 *z2;
+	unsigned int mt_flags = INPUT_MT_DIRECT;
 	unsigned int slots;
 	int cal_size;
 	int error;
@@ -2358,13 +2359,16 @@ static int apple_z2_probe(struct spi_device *spi)
 	input_set_abs_params(z2->input_dev, ABS_MT_ORIENTATION, -32768, 32767,
 			     0, 0);
 
-	if (apple_z2_is_j172(z2))
+	if (apple_z2_is_j172(z2)) {
 		slots = APPLE_Z2_J172_MAX_CONTACTS;
-	else if (apple_z2_is_iphone7_plus(z2))
+	} else if (apple_z2_is_iphone7_plus(z2)) {
 		slots = 10;
-	else
+		if (apple_z2_is_d11(z2))
+			mt_flags |= INPUT_MT_DROP_UNUSED;
+	} else {
 		slots = 256;
-	error = input_mt_init_slots(z2->input_dev, slots, INPUT_MT_DIRECT);
+	}
+	error = input_mt_init_slots(z2->input_dev, slots, mt_flags);
 	if (error)
 		return dev_err_probe(dev, error,
 				     "unable to initialize multitouch slots\n");
